sound/soc/sof/ipc4-topology.c

```diff
@@ -205,6 +214,35 @@ static int sof_ipc4_setup_process_init_byte_kcontrol(struct snd_sof_widget *swid
 		process->init_byte_kcontrol = control_data;
 	}
 
+	module_in_basefw = sof_ipc4_module_is_in_basefw(swidget->module_info);
+	if (!module_in_basefw) {
+		struct sof_ipc4_base_module_cfg_ext *base_ext;
+		int size;
+
+		/*
+		 * Each loadable module must has a byte kcontrol for its instance initialization,
+		 * we are safe without nullable check for control_data here.
+		 */
+		control_data = process->init_byte_kcontrol;
+		base_ext = (struct sof_ipc4_base_module_cfg_ext *)control_data->data->data;
+		if (base_ext->num_sink_pin_fmts < 1 || base_ext->num_source_pin_fmts < 1) {
+			dev_err(scomp->dev, "Invalid number of pin formats in config extension for %s\n",
+				swidget->widget->name);
+			return -EINVAL;
+		}
+
+		size = sizeof(struct sof_ipc4_base_module_cfg_ext) +
+					base_ext->priv_param_len +
+					(base_ext->num_sink_pin_fmts +
+					base_ext->num_source_pin_fmts) *
+					sizeof(struct sof_ipc4_pin_format);
+		if (size != control_data->data->size) {
```

Comment:
I think we should remove this patch and change the line 981 to set the size based on the size calculation here.

Reply:
I add this check just for the correctness of the kcontrol data, but if the kcontrol data is incorrect, the firmware will report the error gracefully, just as you don't suggest to check payload_with_output_fmt for loadable before. so I will remove this check.

> change the line 981 to set the size based on the size calculation here.
I think based on the removal of base module config extension size check, we don't need this anymore. because if a correct blob is provided, the calcualted size should equal control_data->data->size, if a wrong blob is provided, firmware will show error anyway.